### PR TITLE
Migrate dockerfiles to use cmake apt repo

### PR DESCRIPTION
**Pull request by masterleinad**
Cherry-picked from https://github.com/kokkos/kokkos/pull/3913. Given that we constantly running into problems downloading `CMake`, I at least want to discuss the approach of using the `apt` repository provided by `CMake` instead.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Review by crtrott — APPROVED** [review]
Seems fine to me but put on the list for todays discussion.

**Review by dalg24 — CHANGES_REQUESTED** [review]
I like our being able to specify exactly what version of CMake we want.
We would lose that ability with the changes suggested here.

How often does retrieving the GPG keys from the server fail?  They are alternative solution if this is has actually become an issue.
https://github.com/arborx/ArborX/blob/13d7ef8aeea37d6bff3425d6bd9c4cf4d77fa056/docker/Dockerfile#L26-L32

**Comment by masterleinad** [comment]
> How often does retrieving the GPG keys from the server fail?

In my experience, it fails a couple of times whenever we need to rebuild an image which happens without any changes to the Dockerfiles multiple times per week (since we are cleaning up images to avoid running out-of-memory). I'm very much in favor of a solution so that we don't need to request a rebuild in particular if only one build fails because of this.

File: scripts/docker/Dockerfile.clang [hunk 1]
```diff
@@ -1,31 +1,19 @@
 FROM nvidia/cuda:9.2-devel
 
-RUN apt-get update && apt-get install -y \
+ENV DEBIAN_FRONTEND noninteractive
+RUN apt-get update && \
+    apt-get install -y apt-transport-https ca-certificates gnupg software-properties-common wget && \
+    wget -O - https://apt.kitware.com/keys/kitware-archive-latest.asc 2>/dev/null | gpg --dearmor - | tee /etc/apt/trusted.gpg.d/kitware.gpg >/dev/null && \
+    apt-add-repository "deb https://apt.kitware.com/ubuntu/ $(cat /etc/lsb-release | grep DISTRIB_CODENAME | awk -F '=' '{print $NF}') main" && \
```
Comment (dalg24) [inline scripts/docker/Dockerfile.clang:7]: there is no need to `cat` the file
`grep DISTRIB_CODENAME /etc/lsb-release | cut -d= -f2` does what you want

**Comment by dalg24** [comment]
@masterleinad Do you see any issue with the solution I suggested (get key from a keydump hosted on our server)

**Comment by masterleinad** [comment]
No, not really. I can give it a shot.

**Comment by masterleinad** [comment]
Closing in favor of #4066.

**Pull request closed without merging**